test/sql-luatest/show_create_table_test.lua

```diff
@@ -0,0 +1,456 @@
+        s:create_index('i', {type = 'hash', parts = {{'i', 'integer'}}})
+        _, err = box.execute([[SHOW CREATE TABLE "a";]])
+        res = "SQL does not support non-TREE index type"
+        t.assert_equals(err.message, res)
+        s:drop()
+
+        -- Type of the part in PK different from type of the field.
+        s = box.schema.space.create('a', {format = {{'i', 'integer'}}})
+        s:create_index('i', {parts = {{'i', 'unsigned'}}})
+        _, err = box.execute([[SHOW CREATE TABLE "a";]])
+        res = [[SQL does not support indexes with part type not equal to ]]..
+              [[corresponding field type]]
+        t.assert_equals(err.message, res)
+        s:drop()
+
+        -- Collation of the part in PK different from collation of the field.
+        s = box.schema.space.create('a', {format = {{'i', 'string',
+                                          collation = "unicode_ci"}}})
+        s:create_index('i', {parts = {{'i', 'string', collation = "binary"}}})
+        _, err = box.execute([[SHOW CREATE TABLE "a";]])
+        res = [[SQL does not support indexes with part collation not equal ]]..
+              [[to corresponding field collation]]
+        t.assert_equals(err.message, res)
+        s:drop()
+
+        --
+        -- Spaces with an engine other than "memtx" and "vinyl" cannot be
+        -- created with CREATE TABLE.
+        --
+        _, err = box.execute([[SHOW CREATE TABLE "_vspace";]])
+        res = [[SQL does not support engine other than memtx or vinyl]]
+        t.assert_equals(err.message, res)
+
+        -- Make sure the table, field, and PK names are properly escaped.
+        s = box.schema.space.create('"a"', {format = {{'"i', 'integer'}}})
+        s:create_index('i"', {parts = {{'"i', 'integer'}}})
+        ret = box.execute([[SHOW CREATE TABLE """a""";]])
+        res = [[CREATE TABLE """a"""("""i" INTEGER NOT NULL,]]..
+              [[CONSTRAINT "i""" PRIMARY KEY("""i"));]]
+        t.assert_equals(ret.rows[1][1]:gsub('\n', ''), res)
+        s:drop()
+    end)
+end
+
+g.test_field_foreign_key_from_lua = function()
+    g.server:exec(function()
+        local format = {{'i', 'integer'}}
+        box.schema.space.create('a', {format = format})
+
+        -- Working example.
+        format[1].foreign_key = {a = {space = 'a', field = 'i'}}
+        box.schema.space.create('b', {format = format})
+        box.space.b:create_index('i', {parts = {{'i', 'integer'}}})
+        local ret = box.execute([[SHOW CREATE TABLE "b";]])
+        local res = [[CREATE TABLE "b"("i" INTEGER NOT NULL ]]..
+                    [[CONSTRAINT "a" REFERENCES "a"("i"),]]..
+                    [[CONSTRAINT "i" PRIMARY KEY("i"));]]
+        t.assert_equals(ret.rows[1][1]:gsub('\n', ''), res)
+
+        res = [[CREATE TABLE "b"("i" INTEGER NOT NULL,]]..
+              [[CONSTRAINT "i" PRIMARY KEY("i"));]]..
+              [[/* Foreign key "a" cannot be defined in this statement. */]]
```

Comment:
AFAIU, the purpose of `SHOW CREATE TABLE` usually is to export the schema into an SQL script and then re-apply it anew in some other place. And I still see problems with it in the current API:

* If you silently ignore non-printable parts like unknown foreign keys, then you generate a valid but just wrong SQL statement. People will print it into a script, apply, won't see any errors, and then stuff will explode in their tests. Not at the time of the SQL script generation. This is sad. It makes the feature hardly usable as is, I think.

* If a space couldn't be serialized into SQL, as I can see, you still return it as an error-comment. This is fixed for the individual `SHOW CREATE TABLE <name>`, but not for showing all tables. Again, the scenario about schema export will break here. People will need to write boilerplate code around this to check for "statements" which start from `/*` and thus aren't really statements.

I understand the problems though. You want to return the error somehow. So I would suggest to either fail if anything couldn't be serialized, or just skip such spaces from the result (sad, no full schema, but the returned spaces will be at least full and valid), or return 2 fields: `statement` and `error` or `message`. `statement` is always fully serialized space-SQL or `nil`. `error` / `message` is either `nil` or the message. Then people at least will be able to check for individual errors normally, instead of doing `res.rows[i]:starts('/*')` which is just extremely weird. Note, that even this hack wouldn't help against the code block which I commented. When you appended the error somewhere in the end.

You summoned your product-pokemon last time? ;) Now lets do a battle - I summon mine! @Mons, ATTACK! (please)

Reply:
> If you silently ignore non-printable parts like unknown foreign keys, then you generate a valid but just wrong SQL statement.

some things is not provided by `CREATE TABLE` operator by design. For example non-unique indexes, etc.

So the pattern will never work in ALL user scenarios. I think it is a normal situation.
Worse if a user asks the list and receive an error because one item of the list contains non-printable index.

Reply:
I mostly agree with @Gerold103. Explicit error in such a case is much better than silent hidden rake of wrong SQL.
Nevertheless I'd like to have partially dumped statement for such "undumpable" spaces, rather than none at all.
So, I'd expect explicit error in case of some things, that could not be dumped. And additional modifier, that will dump a space anyway, with errors shown as comments (like `show create table <identifier> ignore erorrs`)

Reply:
I discussed with @unera and we decided to introduce 4 versions of `SHOW CREATE TABLE`:
1) `SHOW CREATE TABLE <table_name>;` - return a description of the space. On any problem, it gives an error;
2) `SHOW CREATE TABLE <table_name> INCLUDING ERRORS;` - return a description of the space. If the table does not exist, an error is thrown, any other problems are included as comments to `CREATE TABLE`;
3) `SHOW CREATE TABLE;` - return descriptions of all non-system available spaces that can be described with `SHOW CREATE TABLE <table_name>;`. Spaces that cannot be described with `SHOW CREATE TABLE <table_name>;` are ignored;
4) `SHOW CREATE TABLE INCLUDING ERRORS;` -  return descriptions of all non-system available spaces. Any problems are included as comments to `CREATE TABLE`.

There is one more important change: if the INCLUDING ERRORS clause is used, the description will be returned even if the description is not working `CREATE TABLE`. For example, for a space created with `box.schema.space.create('a')`, the description would be:
```
CREATE TABLE "a";
/* Space "a" cannot be defined in this statement: PRIMARY KEY missing */
/* Space "a" cannot be defined in this statement: SQL does not support space without format */
```

Reply:
>
```sql
CREATE TABLE "a";
/* Space "a" cannot be defined in this statement: PRIMARY KEY missing */
/* Space "a" cannot be defined in this statement: SQL does not support space without format */
```

Could we place comments into brackets?

```sql
CREATE TABLE "a" (
/* Space "a" cannot be defined in this statement: PRIMARY KEY missing */
/* Space "a" cannot be defined in this statement: SQL does not support space without format */
);
```

Reply:
If a table is created, how it can throw an error 'Wrong engine'?

Anyway, we can place the comment inside brackets (As the first or last comment).

Reply:
For example:
```
SHOW CREATE TABLE "_space";
```

Reply:
Why so complicated? Did you all read my comment fully? Citing, I wrote:

"... *or return 2 fields: `statement` and `error` or `message`. `statement` is always a fully serialized space-SQL or `nil`. `error` / `message` is either `nil` or an error message. Then people at least will be able to check for individual errors normally* ...".

What is wrong with this? You get complete error messages + you don't fail the whole `SHOW CREATE TABLE` because of just one space + you don't have incorrect SQL statements. And you don't need to introduce multiple versions of `SHOW` with weird syntax.

Reply:
> Why so complicated?

I think, because in case of minor errors I'd like to receive my SQL statement with some comments, not nil and error message and then write this statement by hands. There are often no ability to fix an issue to make dump successful.

Reply:
I'm not sure if we should return both a not-NULL result and an error in the same answer. Also, there is a use case similar to this case in the downgrade feature.